rsyslog/go/src/github.com/mmexternal/undefined_field/undefined_field.go

```diff
@@ -0,0 +1,239 @@
+	initial_logging_file_path          = "/var/log/rsyslog/rsyslog.log"
+)
+
+var (
+	use_undefined bool
+	keep_fields []string
+	keep_empty_fields []string
+	undefined_name string
+	undefined_to_string bool
+	undefined_dot_replace_char string
+	undefined_max_num_fields int64
+	logfile *os.File
+	noaction          = false
+	replacer          = &strings.Replacer{}
+	basic_keep_fields = "_STREAM_ID,_SYSTEMD_INVOCATION_ID"
+)
+
+func onInit() {
+	// opening the rsyslog log file
+	logging_file_path := initial_logging_file_path
+	if eval := os.Getenv("LOGGING_FILE_PATH"); eval != "" {
+		logging_file_path = eval
+	}
+	if f, err := os.OpenFile(logging_file_path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644); err == nil {
+		logfile = f
+	} else {
+		panic(err.Error())
+	}
+
+	// initializing
+	use_undefined = initial_use_undefined
+	default_keep_fields := initial_default_keep_fields
+	extra_keep_fields := initial_extra_keep_fields
+	undefined_name = initial_undefined_name
+	tmp_keep_empty_fields := initial_keep_empty_fields
+	undefined_to_string = initial_undefined_to_string
+	undefined_dot_replace_char = initial_undefined_dot_replace_char
+	undefined_max_num_fields = initial_undefined_max_num_fields
+
+	config, err := os.Open("/var/lib/rsyslog.pod/undefined.yaml")
+	if err != nil {
+	  panic(err.Error())
+	}
+	defer config.Close()
+
+	reader := bufio.NewReader(config)
+	scanner := bufio.NewScanner(reader)
+	scanner.Split(bufio.ScanLines)
+	for scanner.Scan() {
+		rawStr := scanner.Text()
+		keyVal := strings.Split(rawStr, ": ")
+		if strings.HasPrefix(keyVal[0], "#") || len(keyVal) < 2 {
+			continue
+		}
+		switch keyVal[0] {
+		case "CDM_USE_UNDEFINED":
+			if rval, err := strconv.ParseBool(keyVal[1]); err == nil {
+				use_undefined = rval
+			}
+		case "CDM_DEFAULT_KEEP_FIELDS":
+			default_keep_fields = keyVal[1]
+		case "CDM_EXTRA_KEEP_FIELDS":
+			extra_keep_fields = keyVal[1]
+		case "CDM_UNDEFINED_NAME":
+			undefined_name = keyVal[1]
+		case "CDM_KEEP_EMPTY_FIELDS":
+			tmp_keep_empty_fields = keyVal[1]
+		case "CDM_UNDEFINED_TO_STRING":
+			if rval, err := strconv.ParseBool(keyVal[1]); err == nil {
+				undefined_to_string = rval
+			}
+		case "CDM_UNDEFINED_DOT_REPLACE_CHAR":
+			undefined_dot_replace_char = keyVal[1]
+		case "CDM_UNDEFINED_MAX_NUM_FIELDS":
+			var rval64 int64
+			if rval64, err = strconv.ParseInt(keyVal[1], 10, 64); err == nil {
+				undefined_max_num_fields = rval64
+			}
+		}
+	}
+
+	if !use_undefined && tmp_keep_empty_fields == "" && !undefined_to_string && undefined_dot_replace_char == "UNUSED" {
+		noaction = true
+	}
+	if undefined_max_num_fields == -1 {
+		undefined_max_num_fields = 9223372036854775807
+	}
+	if use_undefined {
+		tmp_basic := strings.Split(basic_keep_fields, ",")
+		tmp_default := strings.Split(default_keep_fields, ",")
+		tmp_extra := strings.Split(extra_keep_fields, ",")
+		keep_fields = append(tmp_basic, tmp_default...)
+		keep_fields = append(keep_fields, tmp_extra...)
```

Comment:
in general it is better to make `keep_fields` a `map[string]string` so you can use `O(1)` hash lookups and avoid having to use the `O(n)` function `contains` - same with `keep_empty_fields`.